### DISCUSSION TITLE
How to print or collect the current running topology during simulation runtime?

### DISCUSSION THREAD

**Discussion by ender0229** [omnetpp/omnetpp#1070] [Q&A] [answered]
I try to collect current topology information when the topology changes while the simulation is running. Are any existing modules implemented for INET 4.5.2? Or how can I do it? Any guidance would be appreciated.

**Comment by rhornig** [comment, marked as answer]
OMNeT++ Assistant suggest:

> To collect topology information during a simulation in OMNeT++ using INET 4.5.2, you can leverage the cTopology class, which provides a way to explore the network topology programmatically. There isn't a specific pre-built module in INET 4.5.2 dedicated to dynamically tracking topology changes, but you can implement this functionality by writing some custom code.
> 
> Here's a general approach you can follow:
> 
> 1. Create a Custom Module: Develop a custom module or modify an existing one to track topology changes. This module should use cTopology to analyze the network.
> 2. Use cTopology Class: The cTopology class can be used to extract various network topology information, such as node connections, paths, etc. You can use methods like extractByProperty() or calculateWeightedSingleShortestPathsTo() for specific topology information.
> 3. Handling Topology Changes: To detect and act upon topology changes, you might want to listen for specific events or signals that indicate a change in the network. This could be a link going down, a node moving (in case of wireless networks), etc. You can use the signal and event handling mechanisms in OMNeT++ for this.
> 4. Periodic Topology Updates: If the network topology changes dynamically (as in mobile ad hoc networks), you may need to periodically reevaluate the topology. This can be done by scheduling self-messages at regular intervals to trigger topology reevaluation.
> 5. Logging or Storing Topology Data: Once you detect a change and reevaluate the topology, you might want to log this information or store it for further analysis. This can be done using OMNeT++'s logging facilities or by writing to a file.
> 6. Integration with INET: Ensure your custom module or changes are well-integrated with INET's network components. Pay attention to INET's existing mechanisms for handling topology-related information.
> 7. Testing and Validation: Test your implementation in various scenarios to ensure it correctly detects and handles topology changes.
> 8. 
> 
> For specific API calls and more detailed instructions, you should consult the OMNeT++ Simulation Manual and INET Framework documentation. The cTopology class and its usage are well documented in the OMNeT++ API documentation, which can be found at [OMNeT++ API - cTopology](https://doc.omnetpp.org/omnetpp/api/classomnetpp_1_1cTopology.html).
> 
> Remember that this approach requires a good understanding of both OMNeT++ and the INET framework, as well as C++ programming skills.

**Reply by rhornig** [reply]
It does not. It's just a utility function that extracts the data. There are various signals in the kernel you can subscribe to (PRE_MODEL_CHANGE, POST_MODEL_CHANGE) that are called by the simulator to notify you about **topology** changes. That being said, you are right the the commands you use (startup/shutdown) do NOT change the topology of the network. They just disables/enables that node, but it is still there and did not disappear. But create-module, delete-module, connect and disconnect cause topology changes.
